docs: DVP update due v1.9.0 (#20647)
deckhouse/deckhouse: docs/site/pages/virtualization-platform/documentation/user/resource-managment/USB_DEVICES.md

---
title: "USB devices"
permalink: en/virtualization-platform/documentation/user/resource-management/usb-devices.html
---

{% alert level="warning" %}

> USB device passthrough is available only in the **Enterprise Edition (EE)** of the Deckhouse Virtualization Platform.

Reviewer: ```suggestion
USB device passthrough is available only in the Deckhouse Virtualization Platform **Enterprise Edition (EE)**.
```

- [USBDevice](/modules/virtualization/cr.html#usbdevice) (namespaced resource) — represents a USB device available for attachment to virtual machines in a given namespace.

## How it works

USB device passthrough follows a defined lifecycle — from device discovery on a node to attachment to a virtual machine:

1. The DRA driver discovers USB devices on cluster nodes and publishes them to the Kubernetes API as ResourceSlices. The module controller creates [NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) resources from that data.

1. An administrator assigns a namespace to the [NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) resource by setting the `.spec.assignedNamespace` resource field. This makes the device available in that namespace.

1. After the namespace is assigned, the module controller automatically creates a corresponding [USBDevice](/modules/virtualization/cr.html#usbdevice) resource in that namespace.

> 1. The [USBDevice](/modules/virtualization/cr.html#usbdevice) is attached to a virtual machine by adding it to the `.spec.usbDevices` resource field of the [VirtualMachine](/modules/virtualization/cr.html#virtualmachine) resource.

Reviewer: ```suggestion
1. The [USBDevice](/modules/virtualization/cr.html#usbdevice) is attached to a virtual machine by adding it to the `.spec.usbDevices` field of the [VirtualMachine](/modules/virtualization/cr.html#virtualmachine) resource.
```

## Quick start

The following steps describe the minimal workflow for attaching a USB device to a virtual machine:

1. Connect the USB device to a cluster node.
1. Verify that a NodeUSBDevice resource has been created:

   ```bash
   d8 k get nodeusbdevice
   ```

1. Assign a namespace to the [NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) by setting the `.spec.assignedNamespace` resource field.

   ```bash
   d8 k apply -f - <<EOF
   apiVersion: virtualization.deckhouse.io/v1alpha2
   kind: NodeUSBDevice
   metadata:
     name: logitech-webcam
   spec:
     assignedNamespace: my-project
   EOF
   ```

1. Verify that a corresponding [USBDevice](/modules/virtualization/cr.html#usbdevice) resource has been created in the target namespace:

   ```bash
   d8 k get usbdevice -n my-project
   ```

> 1. Add the device to the `.spec.usbDevices` resource field of a [VirtualMachine](/modules/virtualization/cr.html#virtualmachine) resource.

Reviewer: ```suggestion
1. Add the device to the `.spec.usbDevices` field of a [VirtualMachine](/modules/virtualization/cr.html#virtualmachine) resource.
```

   ```bash
   d8 k apply -f - <<EOF
   apiVersion: virtualization.deckhouse.io/v1alpha2
   kind: VirtualMachine
   metadata:
     name: linux-vm
   spec:
     # ... other VM settings ...
     usbDevices:
       - name: logitech-webcam
   EOF
   ```

## NodeUSBDevice

[NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) resource reflects the state of a physical USB device detected on a cluster node. It is a cluster-wide resource that represents a physical USB device on a node.

Example of viewing all discovered USB devices:

```bash
d8 k get nodeusbdevice
```

Example output:

<!-- markdownlint-disable MD031 -->
```console

> NAME                 NODE           READY   ASSIGNED   NAMESPACE   AGE

Reviewer: форматирование таблицы немного поехало во всех колонках кроме NAME. Саджестом не получится предлодить, т.к. внутри кодблока

<img width="880" height="147" alt="Image" src="https://github.com/user-attachments/assets/fb431f22-103d-4757-b4c1-4b24c29b0d3b" />

usb-flash-drive     node-1         True    False                  10m
logitech-webcam     node-2         True    True      my-project   15m
```
{: .nowrap-default }
<!-- markdownlint-enable MD031 -->

### NodeUSBDevice conditions

The status of a [NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) resource is represented by a set of conditions that describe its availability and assignment state:

- **Ready**: Indicates whether the device is ready to use.
  - `Ready`: Device is ready to use.
  - `NotReady`: Device exists but is not ready.
  - `NotFound`: Device is absent on the host.

- **Assigned**: Indicates whether a namespace is assigned to the device.
  - `Assigned`: Namespace is assigned and USBDevice resource is created.
  - `Available`: No namespace is assigned for the device.
  - `InProgress`: Device connection to namespace is in progress.

### Assigning a namespace

> Before a USB device can be attached to a virtual machine, it must be exposed to a specific namespace. To make a USB device available in a specific namespace, set the `.spec.assignedNamespace` resource field:

Reviewer: ```suggestion
Before a USB device can be attached to a virtual machine, it must be exposed to a specific namespace. To make a USB device available in a specific namespace, set the `.spec.assignedNamespace` parameter of the NodeUSBDevice resource::
```

```bash
d8 k apply -f - <<EOF
apiVersion: virtualization.deckhouse.io/v1alpha2
kind: NodeUSBDevice
metadata:
  name: logitech-webcam
spec:
  assignedNamespace: my-project
EOF
```

After assigning the namespace, a corresponding [USBDevice](/modules/virtualization/cr.html#usbdevice) resource is automatically created in the specified namespace.

## USBDevice

> When the related [NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) has the `.spec.assignedNamespace` resource field set, a corresponding [USBDevice](/modules/virtualization/cr.html#usbdevice) resource is created in that namespace. It is a namespaced resource that represents a USB device available for attachment to virtual machines within a given namespace.

Reviewer: ```suggestion
When the related [NodeUSBDevice](/modules/virtualization/cr.html#nodeusbdevice) has the `.spec.assignedNamespace` field set, a corresponding [USBDevice](/modules/virtualization/cr.html#usbdevice) resource is created in that namespace. It is a namespaced resource that represents a USB device available for attachment to virtual machines within a given namespace.
```

- `vendorID`: USB vendor ID (hexadecimal format).
- `productID`: USB product ID (hexadecimal format).
- `bus`: USB bus number.
- `deviceNumber`: USB device number on the bus.
- `serial`: Device serial number.
- `manufacturer`: Device manufacturer name.
- `product`: Device product name.
- `name`: Device name.

### USBDevice conditions

The [USBDevice](/modules/virtualization/cr.html#usbdevice) resource provides status conditions that reflect its readiness and attachment state.

- **Ready**: Indicates whether the device is ready to use.
  - `Ready`: Device is ready to use.
  - `NotReady`: Device exists but is not ready.
  - `NotFound`: Device is absent on the host.

- **Attached**: Indicates whether the device is attached to a virtual machine.
  - `AttachedToVirtualMachine`: Device is attached to a VM.
  - `Available`: Device is available for attachment.
  - `NoFreeUSBIPPort`: Device is requested by a VM but cannot be attached because there are no free USBIP ports on the target node. In this case, `Attached=False`.

## Attaching USB device to VM

> After the [USBDevice](/modules/virtualization/cr.html#usbdevice) resource is available in a namespace, it can be attached to a virtual machine. To attach a USB device to a virtual machine, add the device to the `.spec.usbDevices` resource field of the [VirtualMachine](/modules/virtualization/cr.html#virtualmachine) resource specification:

Reviewer: ```suggestion
After the [USBDevice](/modules/virtualization/cr.html#usbdevice) resource is available in a namespace, it can be attached to a virtual machine. To attach a USB device to a virtual machine, add the device to the `.spec.usbDevices` field of the [VirtualMachine](/modules/virtualization/cr.html#virtualmachine) resource specification:
```

```bash
d8 k describe nodeusbdevice <device-name>
```

Example output:

```console
Name:         logitech-webcam
Namespace:
Labels:       <none>
Annotations:  <none>
API Version:  virtualization.deckhouse.io/v1alpha2
Kind:         NodeUSBDevice
Metadata:
  Creation Timestamp:  2024-01-15T10:30:00Z
  Generation:          1
  UID:                 abc123-def456-ghi789
Spec:
  Assigned Namespace:  my-project
Status:
  Node Name:           node-2
  Attributes:
    Bus:               1
    Device Number:     2
    Manufacturer:      Logitech
    Name:              Webcam C920
    Product:           Webcam C920
    Product ID:        082d
    Serial:            ABC123456
    Vendor ID:         046d
  Conditions:
    Type:              Ready
    Status:            True
    Reason:            Ready
    Message:           Device is ready to use
    Type:              Assigned
    Status:            True
    Reason:            Assigned
    Message:           Namespace is assigned for the device
  Observed Generation: 1
```

{% alert level="info" %}
If a USB device is physically disconnected from the node, the `Attached` condition becomes `False`.  

> Both `USBDevice` and `NodeUSBDevice` resources update their status conditions to indicate that the device is no longer present on the host.

Reviewer: ```suggestion
Both USBDevice and NodeUSBDevice resources update their status conditions to indicate that the device is no longer present on the host.
```
